shells/android/java/arcs/crdt/parcelables/ParcelableCrdtEntity.kt

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright 2019 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.crdt.parcelables
+
+import android.os.Parcel
+import android.os.Parcelable
+import arcs.common.Referencable
+import arcs.crdt.CrdtEntity
+import arcs.crdt.CrdtSet
+import arcs.crdt.CrdtSingleton
+import arcs.data.FieldName
+
+/** Container of [Parcelable] implementations for the data and ops classes of [CrdtEntity]. */
+object ParcelableCrdtEntity {
+    /** Parcelable variant of [CrdtEntity.ReferenceImpl]. */
+    data class ReferenceImpl(
+        override val actual: CrdtEntity.ReferenceImpl
+    ) : ParcelableReferencable {
+        override fun writeToParcel(parcel: Parcel, flags: Int) {
+            super.writeToParcel(parcel, flags)
+            parcel.writeString(actual.id)
+        }
+
+        override fun describeContents(): Int = 0
+
+        /** Don't use this directly; use the [ParcelableReferencable] base class instead. */
+        internal companion object CREATOR : Parcelable.Creator<ReferenceImpl> {
+            override fun createFromParcel(parcel: Parcel): ReferenceImpl {
+                val id = requireNotNull(parcel.readString(), {
+                    "ID not found in parcel when reading ParcelableCrdtEntity.ReferenceImpl"
+                })
+                return ReferenceImpl(CrdtEntity.ReferenceImpl(id))
+            }
+
+            override fun newArray(size: Int): Array<ReferenceImpl?> = arrayOfNulls(size)
+        }
+    }
+
+
+    /** Parcelable variant of [CrdtEntity.Data]. */
+    data class Data(
+        override val actual: CrdtEntity.Data
+    ) : ParcelableCrdtData<CrdtEntity.Data> {
+        override var versionMap = actual.versionMap
+
+        @Suppress("UNCHECKED_CAST")
+        override fun writeToParcel(parcel: Parcel, flags: Int) {
+            parcel.writeTypedObject(actual.versionMap.toParcelable(), flags)
+
+            parcel.writeInt(actual.singletons.size)
+            actual.singletons.forEach { (field, value) ->
+                parcel.writeString(field)
+                parcel.writeTypedObject(ParcelableCrdtSingleton.Data(
+                    value.data as CrdtSingleton.Data<Referencable>), flags
+                )
+            }
+
+            parcel.writeInt(actual.collections.size)
+            actual.collections.forEach { (field, value) ->
+                parcel.writeString(field)
+                parcel.writeTypedObject(
+                    ParcelableCrdtSet.Data(value.data as CrdtSet.Data<Referencable>), flags)
```

Comment:
Is there a `toParcelable` extension function for CrdtSet.Data that you could use here?

Reply:
Done